### PR TITLE
Resolve command paths and set UTF-8 mode for tests on Windows

### PR DESCRIPTION
**Pull request by dagewa**
#777 assumed all DIALS command paths end with the extension `.bat` on Windows.

Not so, in some circumstances (cmake build?) the commands are actually `.exe`. Procrunner always did this properly, so let's borrow the `windows_resolve` function from there.

First step in trying to deal with #818

### CODE REVIEW

**Comment by dagewa** [comment]
@Anthchirp, this copies your function from procrunner, with a note in the docstring. Is this attribution sufficient?

**Comment by ndevenish** [comment]
Does `shutil.which` work? We've started using it elsewhere and it looks like it should "do the right thing" in [the documentation](https://docs.python.org/3/library/shutil.html#shutil.which).

**Comment by graeme-winter** [comment]
> Does `shutil.which` work? We've started using it elsewhere and it looks like it should "do the right thing" in [the documentation](https://docs.python.org/3/library/shutil.html#shutil.which).

I note we use it in the tests e.g. 

```
def test_from_image_files_implicit_chunks(dials_data, tmp_path):
    image_files = sorted(
        dials_data("centroid_test_data", pathlib=True).glob("centroid*.cbf")
    )

    # Import from the image files
    result = subprocess.run(
        [
            shutil.which("dials.import"),
            "output.experiments=imported.expt",
            "split=3",
        ]
        + image_files,
        cwd=tmp_path,
        capture_output=True,
    )
    assert not result.returncode
    assert (tmp_path / "imported.expt").is_file()
    exp = load.experiment_list(tmp_path / "imported.expt")
    assert len(exp) == 3
```

**Comment by ndevenish** [comment]
I'm left more confused; it looks like `windows_resolve` is using `shutil.which`... with special handling if you have passed in the full path already. Wouldn't you _not_ need to find the executable in that case?

**Comment by dagewa** [comment]
Good point, I barely looked at `windows_resolve`, just copied it in because I knew this worked in procrunner. Happy to change this to a bare `shutil.which`, though I suppose we really need to do this _everywhere_

**Comment by dagewa** [comment]
Ok, the special handling is so you can pass `C:\\some\\path\\dials.thing` and it would resolve to the `.exe`, `.bat` or whatever, whereas `shutil.which` on its own won't, apparently. We probably don't need that special handling in usage across DIALS, xia2 and so on

**Comment by dagewa** [comment]
Ok, switched to `shutil.which`, but pre-commits showed a mypy type issue: `shutil.which` might return `None`. So far I only fixed that in the three places it complained about, where functions had type annotations.

**Comment by dagewa** [comment]
All the other places are tests, so I think this is fine

File: src/xia2/Driver/DriverHelper.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 
+import logging
 import os
 import random
 import signal
 import stat
 import string
 
+logger = logging.getLogger("xia2.Driver.DriverHelper")
+
```
Comment (ndevenish) [inline src/xia2/Driver/DriverHelper.py]: This is leftover from old implementation?
```suggestion
import os
import random
import signal
import stat
import string

```

**Comment by ndevenish** [comment]
I think I'd be tempted to reduce the `shutil.which` failure handlings to plain `assert`. It's a pity there is no builtin for some equivalent of `shutil.which("xia2") or raise(RuntimeError("No xia2"))`

File: src/xia2/Driver/DriverHelper.py (hunk 1 above)
Comment (dagewa) [reply src/xia2/Driver/DriverHelper.py]: Yes, it was residual from `windows_resolve`

**Comment by dagewa** [comment]
```
assert (cmd := shutil.which("dials.import"))
```
?

**Comment by ndevenish** [comment]
FWIW assigning inline in the assert will break if python is ever run with PYTHONOPTIMIZE / `-O`

```
% python3 -O
Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:54:21) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> assert (a := 3)
>>> a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'a' is not defined
>>>
```

**Pull request merged**